File: lib/checkout.ts
import {
  InttegroAPIError,
  InttegroClient,
  type CreateOrderRequest,
  type Product,
} from '@inttegro/inttegro-sdk';

/**
 * Inttegro integration map
 *
 * INTTEGRO:FLOW This module is server-only. It validates an untrusted request,
 * creates a finalized Order, and returns either its hosted checkout URL (web)
 * or its ID (native payment sheet).
 * INTTEGRO:SECURITY [server-api-key] INTTEGRO_API_KEY and the demo Customer ID
 * must never move into a NEXT_PUBLIC_* variable, Client Component, or mobile
 * response. A browser or application bundle cannot keep a merchant secret.
 * INTTEGRO:DECISION [hosted-checkout] Web customers use the hosted invoice URL
 * returned by Inttegro, keeping payment entry and provider confirmation out of
 * the merchant application.
 * INTTEGRO:ALTERNATIVE [hosted-checkout] Direct API payment is appropriate only
 * when the merchant intends to own the larger payment state machine, recovery
 * UI, method-specific behavior, testing, and compliance analysis.
 * INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
 * INTTEGRO:DOCS https://studio.inttegro.com/orders
 * INTTEGRO:DOCS https://studio.inttegro.com/keys
 * See ../../INTEGRATION_GUIDE.md and ../../integration-decisions.json for the complete
 * rationale and machine-readable alternatives.
 */

export type CheckoutInput = {
  name: string;
  email: string;
  phone: string;
  attemptId: string;
};

export type MobileCheckoutInput = {
  attemptId: string;
};

export type CatalogSelection = {
  type: Product['type'];
  name: string;
  about?: string;
  reference?: string;
  price: NonNullable<Product['prices']>[number]['nominal'];
};

type MobileCreateOrderRequest = CreateOrderRequest & {
  payment_method_types: ['mobile_money'];
};

export class DemoError extends Error {
  readonly code: 'configuration_error' | 'validation_error' | 'api_error';

  constructor(
    code: 'configuration_error' | 'validation_error' | 'api_error',
    message: string,
  ) {
    super(message);
    this.code = code;
  }
}

const emailPattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
const attemptPattern = /^[A-Za-z0-9_-]{8,100}$/;

export function parseCheckoutInput(form: FormData): CheckoutInput {
  const input = {
    name: String(form.get('name') ?? '').trim(),
    email: String(form.get('email') ?? '').trim(),
    phone: String(form.get('phone') ?? '').trim(),
    attemptId: String(form.get('attempt_id') ?? '').trim(),
  };

  if (!input.name || !input.phone || !emailPattern.test(input.email)) {
    throw new DemoError('validation_error', 'Enter a name, valid email, and phone number.');
  }
  if (!attemptPattern.test(input.attemptId)) {
    throw new DemoError('validation_error', 'Start a fresh checkout and try again.');
  }
  return input;
}

export function parseMobileCheckoutInput(value: unknown): MobileCheckoutInput {
  if (
    typeof value !== 'object'
    || value === null
    || Array.isArray(value)
    || Object.keys(value).some((key) => key !== 'attemptId')
    || !('attemptId' in value)
    || typeof value.attemptId !== 'string'
  ) {
    throw new DemoError('validation_error', 'Send a valid mobile checkout attempt ID.');
  }

  const attemptId = value.attemptId.trim();
  if (!attemptPattern.test(attemptId)) {
    throw new DemoError('validation_error', 'Start a fresh mobile checkout and try again.');
  }
  return { attemptId };
}

function configuredOrigin(requestOrigin: string): string {
  // INTTEGRO:SECURITY [configured-public-origin] A configured public origin is
  // safer in production. The request origin keeps localhost convenient, but it
  // is safe only when Host/forwarded headers come through a trusted proxy chain.
  // Inttegro return URLs must be absolute HTTP(S) URLs and must not already
  // contain order_id; Inttegro appends the created order ID on return.
  const value = process.env.INTTEGRO_DEMO_PUBLIC_URL?.trim() || requestOrigin;
  try {
    return new URL(value).origin;
  } catch {
    throw new DemoError('configuration_error', 'The demo public URL is invalid.');
  }
}

export function selectCatalogProduct(product: Product, priceId: string): CatalogSelection {
  // INTTEGRO:SECURITY [catalog-authority] The browser never submits product or
  // price data. The trusted server resolves both IDs and validates the selected
  // price against the Product returned by Inttegro before creating the Order.
  // This prevents a customer from changing the amount or swapping catalogue
  // records in an untrusted form or mobile request.
  if (!product.active) {
    throw new DemoError('configuration_error', 'The configured demo product is not active.');
  }
  const price = product.prices?.find((candidate) => candidate.id === priceId && candidate.active);
  if (!price || !Number.isSafeInteger(price.nominal.value) || price.nominal.value <= 0) {
    throw new DemoError('configuration_error', 'The configured demo price is not active for this product.');
  }
  return {
    type: product.type,
    name: product.name,
    ...(product.about ? { about: product.about } : {}),
    ...(product.reference ? { reference: product.reference } : {}),
    price: price.nominal,
  };
}

function configuredCatalogIds(): { productId: string; priceId: string } {
  const productId = process.env.INTTEGRO_DEMO_PRODUCT_ID?.trim() || '';
  const priceId = process.env.INTTEGRO_DEMO_PRICE_ID?.trim() || '';
  if (!/^prod_[A-Za-z0-9]+$/.test(productId) || !/^pr_[A-Za-z0-9]+$/.test(priceId)) {
    throw new DemoError(
      'configuration_error',
      'Set INTTEGRO_DEMO_PRODUCT_ID and INTTEGRO_DEMO_PRICE_ID on the server.',
    );
  }
  return { productId, priceId };
}

async function loadCatalogSelection(inttegro: InttegroClient): Promise<CatalogSelection> {
  const { productId, priceId } = configuredCatalogIds();
  // INTTEGRO:FLOW [catalog-lookup] Fetch the Product at checkout time so the
  // Order uses current server-authoritative catalogue data. High-volume apps
  // can cache this read briefly, but must define invalidation for product,
  // publication, and price changes instead of keeping an unbounded process cache.
  // https://studio.inttegro.com/products
  const product = await inttegro.products.lookup({ product_id: productId });
  return selectCatalogProduct(product, priceId);
}

export function buildOrderRequest(
  input: CheckoutInput,
  origin: string,
  product: CatalogSelection,
): CreateOrderRequest {
  return {
    // INTTEGRO:DECISION [stable-idempotency-key] One browser-rendered attempt ID
    // represents one logical order creation. Reuse it for retries. Production
    // systems should persist a key derived from a durable cart or invoice ID;
    // generating a fresh value on every retry defeats replay protection.
    // https://studio.inttegro.com/idempotency
    request_meta: { idempotency_key: `demo-${input.attemptId}` },

    // INTTEGRO:DECISION [inline-customer] This is a guest checkout, so the
    // request carries customer_data. Account-based products should resolve a
    // stored Customer server-side and send customer_id instead—never both.
    customer_data: {
      name: input.name,
      email_address: input.email,
      phone_number: input.phone,
    },

    // INTTEGRO:DECISION [finalize-on-create] The demo's lines and total are
    // already settled. finalize freezes the order and generates its invoice
    // formats in the same operation. If shipping, tax, approval, or inventory
    // can still change, create a draft, update it, then finalize explicitly.
    finalize: true,
    checkout_settings: {
      redirect_url: `${origin}/complete`,
      cancel_url: `${origin}/cancel`,
    },
    line_items: [
      {
        type: 'product',
        product: {
          // INTTEGRO:DECISION [catalog-snapshot] This cross-SDK demo first looks
          // up an Inttegro Product and its configured Price, then snapshots the
          // verified fields as an inline Order item. That keeps the example
          // portable across SDK releases whose typed order models do not yet all
          // expose the catalog-reference union.
          // INTTEGRO:ALTERNATIVE [catalog-snapshot] When your SDK exposes it,
          // send { product_id, price_id, quantity } instead. That is the most
          // compact catalogue-backed shape; never mix it with inline fields.
          ...product,
          quantity: 1,
        },
      },
    ],
  };
}

export function buildMobileOrderRequest(
  input: MobileCheckoutInput,
  customerId: string,
  product: CatalogSelection,
): MobileCreateOrderRequest {
  return {
    // INTTEGRO:DECISION [stable-idempotency-key] A retry of the same native
    // checkout intent must send the same attempt ID. In production, persist it
    // with the authenticated customer's cart.
    request_meta: { idempotency_key: `mobile-demo-${input.attemptId}` },

    // INTTEGRO:SECURITY [server-customer-id] Customer ownership is resolved by
    // the trusted backend. Never let an untrusted app select an arbitrary
    // customer_id; derive it from the authenticated merchant principal.
    customer_id: customerId,
    // INTTEGRO:DECISION [finalize-on-create] The mobile SDK needs the ID of an
    // immutable checkout-ready order, so this fixed cart finalizes immediately.
    finalize: true,
    // INTTEGRO:DECISION [mobile-backend-boundary] The backend constrains allowed
    // methods and commercial data. The app submits only an opaque attempt ID.
    payment_method_types: ['mobile_money'],
    line_items: [
      {
        type: 'product',
        product: {
          ...product,
          quantity: 1,
        },
      },
    ],
  };
}

export async function createHostedCheckout(input: CheckoutInput, requestOrigin: string) {
  // INTTEGRO:SECURITY [server-api-key] This function must remain on the server.
  // Use the deployment platform's secret manager in production and rotate the
  // key there; never serialize it into an error or response.
  const apiKey = process.env.INTTEGRO_API_KEY?.trim();
  if (!apiKey) {
    throw new DemoError('configuration_error', 'Set INTTEGRO_API_KEY on the server.');
  }

  const origin = configuredOrigin(requestOrigin);
  // INTTEGRO:ALTERNATIVE [server-api-key] A production service may construct and
  // inject one client at startup for fail-fast configuration, connection reuse,
  // and application-owned OpenTelemetry. Per-call construction is clearer here.
  // https://studio.inttegro.com/sdk-observability
  const inttegro = new InttegroClient({ apiKey });

  try {
    const product = await loadCatalogSelection(inttegro);
    const order = await inttegro.orders.create(buildOrderRequest(input, origin, product));

    // INTTEGRO:DECISION [returned-checkout-url] Use the response's URL. Building
    // a URL from order.id would depend on undocumented Inttegro routing details.
    const checkoutUrl = order.invoice?.format?.web?.url;
    if (!checkoutUrl) {
      throw new DemoError('api_error', 'Inttegro did not return a hosted checkout URL.');
    }
    return { checkoutUrl, orderId: order.id };
  } catch (error) {
    // INTTEGRO:SECURITY [safe-error-boundary] Public errors stay stable and
    // non-sensitive. Capture request IDs and safe error metadata in protected
    // server telemetry; never return raw upstream payloads or authorization data.
    if (error instanceof DemoError) throw error;
    if (error instanceof InttegroAPIError) {
      throw new DemoError('api_error', 'Inttegro rejected the checkout request.');
    }
    throw new DemoError('api_error', 'Checkout is temporarily unavailable.');
  }
}

export async function createMobileCheckout(input: MobileCheckoutInput) {
  // INTTEGRO:FLOW [mobile-backend-boundary] This is the illustrative merchant
  // backend for all four native demos. A production route also authenticates
  // the caller, authorizes the cart, rate-limits abuse, and recomputes totals.
  // https://studio.inttegro.com/orders
  const apiKey = process.env.INTTEGRO_API_KEY?.trim();
  const customerId = process.env.INTTEGRO_DEMO_CUSTOMER_ID?.trim();
  if (!apiKey || !customerId) {
    throw new DemoError(
      'configuration_error',
      'Set INTTEGRO_API_KEY and INTTEGRO_DEMO_CUSTOMER_ID on the server.',
    );
  }

  const inttegro = new InttegroClient({ apiKey });
  try {
    const product = await loadCatalogSelection(inttegro);
    const order = await inttegro.orders.create(buildMobileOrderRequest(input, customerId, product));
    if (!order.id?.trim()) {
      throw new DemoError('api_error', 'Inttegro did not return a checkout order ID.');
    }
    // Return only the minimal capability the SDK needs. Product, price,
    // customer_id, payment-method policy, and API key remain server-owned.
    return { orderId: order.id.trim() };
  } catch (error) {
    if (error instanceof DemoError) throw error;
    if (error instanceof InttegroAPIError) {
      throw new DemoError('api_error', 'Inttegro rejected the mobile checkout request.');
    }
    throw new DemoError('api_error', 'Mobile checkout is temporarily unavailable.');
  }
}

File: app/checkout.py
import os
import re
from dataclasses import dataclass
from urllib.parse import urlparse

import inttegro
from inttegro import APIError, Currency, LineItemType, ProductType


# Inttegro integration map
#
# INTTEGRO:FLOW [hosted-checkout] This server-only module validates an untrusted
# ticket request, creates and finalizes an Order, and returns the hosted invoice
# URL used by the FastAPI route.
# INTTEGRO:SECURITY [server-api-key] INTTEGRO_API_KEY stays in the trusted Python
# process. Never publish it through frontend configuration, HTML, client code,
# mobile bundles, or error responses.
# INTTEGRO:ALTERNATIVE [hosted-checkout] Direct API payment offers additional UI
# control but makes the merchant own more payment states, recovery behavior,
# method-specific logic, testing, and compliance analysis.
# INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
# INTTEGRO:DOCS https://studio.inttegro.com/orders
# INTTEGRO:DOCS https://studio.inttegro.com/keys
# See ../../INTEGRATION_GUIDE.md and ../../integration-decisions.json for the full
# rationale and machine-readable alternatives.


class DemoError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message


@dataclass(frozen=True)
class CheckoutInput:
    name: str
    email: str
    phone: str
    attempt_id: str


def parse_checkout_input(name: str, email: str, phone: str, attempt_id: str) -> CheckoutInput:
    value = CheckoutInput(name.strip(), email.strip(), phone.strip(), attempt_id.strip())
    if not value.name or not value.phone or not re.fullmatch(r"[^\s@]+@[^\s@]+\.[^\s@]+", value.email):
        raise DemoError("validation_error", "Enter a name, valid email, and phone number.")
    if not re.fullmatch(r"[A-Za-z0-9_-]{8,100}", value.attempt_id):
        raise DemoError("validation_error", "Start a fresh checkout and try again.")
    return value


def validated_origin(value: str) -> str:
    # INTTEGRO:SECURITY [configured-public-origin] Production should use an
    # allow-listed configured origin. A request-derived fallback is safe only
    # behind a precisely configured trusted proxy chain. Return URLs must be
    # absolute HTTP(S); Inttegro appends order_id after checkout.
    parsed = urlparse(value)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise DemoError("configuration_error", "The demo public URL is invalid.")
    return f"{parsed.scheme}://{parsed.netloc}"


def build_order_request(value: CheckoutInput, origin: str) -> inttegro.orders.CreateRequest:
    return inttegro.orders.CreateRequest(
        # INTTEGRO:DECISION [stable-idempotency-key] Reuse the key for retries of
        # this logical checkout. Production systems persist a cart- or booking-
        # derived key; creating a fresh key after a timeout permits duplicates.
        # https://studio.inttegro.com/idempotency
        request_meta=inttegro.orders.RequestMeta(idempotency_key=f"demo-{value.attempt_id}"),
        # INTTEGRO:DECISION [inline-customer] customer_data matches this guest
        # flow. Account-based products should resolve customer_id on the trusted
        # server. The Orders API accepts exactly one customer representation.
        customer_data=inttegro.orders.Customer(name=value.name, email_address=value.email, phone_number=value.phone),
        # INTTEGRO:DECISION [finalize-on-create] The item and total are settled,
        # so finalization freezes the order and generates invoice formats in one
        # call. Use draft -> update -> finalize for mutable tax, inventory,
        # approval, shipping, or line items.
        finalize=True,
        checkout_settings=inttegro.orders.CheckoutSettings(redirect_url=f"{origin}/complete", cancel_url=f"{origin}/cancel"),
        line_items=[inttegro.orders.ProductLineItem(
            type=LineItemType.PRODUCT,
            product=inttegro.orders.Product(
                # INTTEGRO:DECISION [inline-product] Inline data keeps this event
                # self-contained. Catalog users can send product_id with price or
                # price_id; do not mix catalog and inline product fields.
                type=ProductType.DIGITAL,
                name="Afterglow Sessions - Courtyard admission",
                quantity=1,
                # INTTEGRO:DECISION [minor-unit-money] 5000 minor units is
                # GHS 50.00. Convert variable input using decimal/money types.
                price=inttegro.PriceParams(currency=Currency.GHS, value=5000),
            ),
        )],
    )


def create_hosted_checkout(value: CheckoutInput, origin: str) -> tuple[str, str]:
    # INTTEGRO:SECURITY [server-api-key] Source this value from a managed runtime
    # secret in production; never include it in logs or client-visible errors.
    api_key = os.environ.get("INTTEGRO_API_KEY", "").strip()
    if not api_key:
        raise DemoError("configuration_error", "Set INTTEGRO_API_KEY on the server.")
    try:
        # INTTEGRO:ALTERNATIVE [server-api-key] Long-running apps can inject one
        # startup-configured client for connection reuse and application-owned
        # OpenTelemetry. The SDK chooses no exporter or vendor.
        # https://studio.inttegro.com/sdk-observability
        order = inttegro.InttegroClient(api_key=api_key).orders.create(build_order_request(value, validated_origin(origin)))
        # INTTEGRO:DECISION [returned-checkout-url] Use the response field rather
        # than constructing a URL from order.id and undocumented routing.
        checkout_url = order.invoice.format.web.url
        if not checkout_url:
            raise DemoError("api_error", "Inttegro did not return a hosted checkout URL.")
        return order.id, checkout_url
    except DemoError:
        raise
    except APIError as error:
        # INTTEGRO:SECURITY [safe-error-boundary] Keep raw upstream payloads,
        # credentials, and traces private. Log bounded metadata and request IDs;
        # expose only stable public categories.
        raise DemoError("api_error", "Inttegro rejected the checkout request.") from error
    except Exception as error:
        raise DemoError("api_error", "Checkout is temporarily unavailable.") from error
